### DISCUSSION TITLE
[Bug] Tabs in sidebar are hidden after exiting some fullscreen videos

### DISCUSSION THREAD

**Discussion by BrontoRed** [piroor/treestyletab#3407] [Q&A]
## Abstract
["Auto-hide sidebar when fullscreen #1548" snippet](https://github.com/piroor/treestyletab/wiki/Code-snippets-for-custom-style-rules#auto-hide-sidebar-when-fullscreen-1548) hides open tabs in sidebar after exiting fullscreen videos. Tabs in sidebar are shown again after hiding and opening sidebar.
Not all fullscreen videos trigger this issue.

## Steps to reproduce

 1. Start Firefox with clean profile.
 2. Install TST.
 3. Enable userChrome.css and "Auto-hide sidebar when fullscreen" snippet as described [here](https://github.com/piroor/treestyletab/wiki/Code-snippets-for-custom-style-rules#for-userchromecss) and [here](https://github.com/piroor/treestyletab/wiki/Code-snippets-for-custom-style-rules#auto-hide-sidebar-when-fullscreen-1548) and restart Firefox
 4. Open [link 1](https://yandex.com/video/preview/1624295599852696422), [link 2](https://yandex.com/video/preview/10487282116056004257) and [link 3](https://www.yahoo.com/entertainment/ahsoka-rosario-dawson-interview-mom-star-wars-debut-the-mandalorian-160056348.html), so that you have 3 tabs
 5. Go to fullscreen video (double left click on video or press fullscreen button in player, but not with F11) in one of the pages above and then back to normal page view

## Expected result
All open tabs in TST are visible in sidebar

## Actual result
All open tabs in TST are hidden. Double tap F1 (or toolbar icon of TST) to restore tabs.

## Environment

 * Platform (OS): Windows 10 Pro 22H2
 * Version of Firefox: 119.0
 * Version (or revision) of Tree Style Tab: 3.9.17 (2023.9.8)

**Comment by piroor** [comment]
userChrome.css customization is out of scope of the issues tracker so I've moved this to discussions.

**Comment by irvinm** [comment]
@piroor I took a little time with this today and was able to reproduce.  The strange thing is that the CSS for #sidebar-box after coming out of the youtube.com video fullscreen looks to be the same as it was before going to fullscreen when it was working.  In addition, although the tabs are not being displayed, they are actually there (I can click around in the sidebar and change the tabs) *AND* if you hide and re-display the sidebar via F1, the tabs are displayed properly again.

Is there any type of paint notification sent to sidebar addons when the browser comes out of fullscreen?  Scratch that, I tested native Firefox sidebar items (history, bookmarks, etc.) and they work properly.  I loaded another sidebar addon (Sideberry) and it shows the same behavior as TST in this situation where it doesn't show, but hiding and displaying it again works fine.  This really feels like a Firefox problem.

Can you think of anything else?

CSS being used:

```
#main-window[inFullscreen] #sidebar-box,
#main-window[inFullscreen] #sidebar-splitter {
  display: none !important;
  width: 0px !important;
}
```

Working view:

![image](https://github.com/piroor/treestyletab/assets/979729/ebb2693a-5ac8-4b5a-8ae7-300d4e7d52d4)

After allowing youtube.com to go fullscreen and return to normal with tabs actually working but not being displayed.  (F1 twice will fix it)

![image](https://github.com/piroor/treestyletab/assets/979729/ef81f39a-a3ae-492d-89d8-f3e52a226c19)

**Reply by piroor** [reply]
The "inFullscreen" attribute of the root element is changed by Firefox and corresponding codes are here:
https://searchfox.org/mozilla-central/rev/dded349c760542c296116a3077695eafabf5fd1f/browser/base/content/browser-fullScreenAndPointerLock.js#312
https://searchfox.org/mozilla-central/rev/dded349c760542c296116a3077695eafabf5fd1f/browser/base/content/browser-fullScreenAndPointerLock.js#354
https://searchfox.org/mozilla-central/rev/dded349c760542c296116a3077695eafabf5fd1f/browser/base/content/browser-fullScreenAndPointerLock.js#366
They remove "inFullscreen" attribute on exiting the fullscreen mode and it should trigger repainting of subframe contents, but it looks fail on these situations.

One possible buggy point is: nested subframes. Firefox's native sidebar panels (Bookmarks and History) are loaded into the sidebar frame directly. On the other hand, WebExtensions based addon's sidebar page is loaded to a subframe in an native sidebar panel abd the panel is loaded into the sidebar frame. This kind indirect loading of subframe contents may be fragile around this kind problem.

> Is there any type of paint notification sent to sidebar addons when the browser comes out of fullscreen?

TST is designed to re-render sidebar contents when the window size is changed, but the event listener can handle only events fired on the sidebar subframe but not on the top level widow of Firefox itself. Resize event looks to be not fired on this situation, maybe because hidden sidebar subframe keeps its size while entering and exiting fullscreen mode.

Sadly `browser.windows.onXXXXX` entry point of WebExtensions API is not available for updating of window size: no such event is notified for windows. So I cannot imagine how to track window size changing. I think I need to give up finding workaroud on TST side...

**Comment by seqizz** [comment]
Just my 2 cents. I had a non-clever hack, seems to work so far:
```
#main-window[inFullscreen] #sidebar-box,
#main-window[inFullscreen] #sidebar-splitter {
  width: 1px !important;
}
```

**Reply by knoajp** [reply]
I have to add min-width on the current Firefox. Maybe not only for me. 🙂
```
#main-window[inFullscreen] #sidebar-box,
#main-window[inFullscreen] #sidebar-splitter {
  width: 1px !important;
  min-width: 1px !important;
}
```
